src/Chapter2.hs

```diff
@@ -841,7 +878,9 @@ list.
 
 🕯 HINT: Use the 'cycle' function
 -}
-rotate = error "rotate: Not implemented!"
+rotate :: Int -> [a] -> [a]
+rotate n xs = if n < 0 then [] else
+  take (length xs) . drop n . cycle $ xs
```

Comment:
Unfortunately, `cycle` fails in runtime when you give it an empty list. So you need to handle this case separately...

Reply:
Interesting. I haven't had much practice working with infinite structures. How would I handle such a case? How is it even possible, given that there is no "end" of the list to wrap back around to be the beginning?

Reply:
I dont think this is actually true, based on my testing. `rotate _ []` returns `[]`. I think probably because of the `take (length xs) <=> take 0` in this case, so lazy evaluation helps us out. Does that seem right to you? When you try it in GHCI do you see different results?

Reply:
Another example:

`take 0 [error "foo"]` => `0` and does not raise an error

Reply:
@chshersh

Reply:
@OxSon Indeed, due to Haskell laziness and how `take` is implemented, your implementation works correctly! My bad, sometimes laziness hard to spot on :slightly_smiling_face: